### ISSUE TITLE
NFS SUBDIR: FATAL:  data directory \"/var/lib/postgresql/data/pgdata\" has wrong ownership

### ISSUE THREAD

**Issue by gentoorax** [cloudnative-pg/cloudnative-pg#2573]
When using the nfs external subdir provisioner for k8s storage, I receive the following error, this works when using longhorn for storage.

I believe this is could be related to issue #625. 

Installed via kubectl apply -f https://raw.githubusercontent.com/cloudnative-pg/cloudnative-pg/release-1.20/releases/cnpg-1.20.2.yaml

cnpg
Version: 1.20.2

nfs-subdir-external-provisioner
Version: 4.0.18

I suspect the NFS subdir mount user or server side NFS user mapping could be at play, but I doubt there's any to stop the init container performing chown/chmod this during init, I've tested I can do this myself with a test pod connected to the NFS PVC.

I've tried a few different storage settings, the NFS subdir provisioner is the default storage class.

Deployment Yaml
```yaml
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: test-cluster
  namespace: psql-alpha
spec:
  instances: 3

  bootstrap:
    initdb:
      database: testdb
      owner: testdbuser
      secret:
        name: testdbuser

  superuserSecret:
    name: test-cluster-superuser

  # Example of rolling update strategy:
  # - unsupervised: automated update of the primary once all
  #                 replicas have been upgraded (default)
  # - supervised: requires manual supervision to perform
  #               the switchover of the primary
  primaryUpdateStrategy: unsupervised

  storage:
    storageClass: nfs-subdir1
    size: 20Gi
```

Error Logs
```text
Container: initdb
Filter
Disconnected
{"level":"info","ts":"2023-08-12T15:01:12Z","msg":"Creating new data directory","logging_pod":"test-cluster-1-initdb","pgdata":"/var/lib/postgresql/data/pgdata","initDbOptions":["--username","postgres","-D","/var/lib/postgresql/data/pgdata","--encoding=UTF8","--lc-collate=C","--lc-ctype=C"]}
{"level":"info","ts":"2023-08-12T15:01:12Z","logger":"initdb","msg":"The files belonging to this database system will be owned by user \"postgres\".\nThis user must also own the server process.\n\nThe database cluster will be initialized with this locale configuration:\n  provider:    libc\n  LC_COLLATE:  C\n  LC_CTYPE:    C\n  LC_MESSAGES: en_US.utf8\n  LC_MONETARY: en_US.utf8\n  LC_NUMERIC:  en_US.utf8\n  LC_TIME:     en_US.utf8\nThe default text search configuration will be set to \"english\".\n\nData page checksums are disabled.\n\ncreating directory /var/lib/postgresql/data/pgdata ... ok\ncreating subdirectories ... ok\nselecting dynamic shared memory implementation ... posix\nselecting default max_connections ... 20\nselecting default shared_buffers ... 400kB\nselecting default time zone ... Etc/UTC\ncreating configuration files ... ok\nrunning bootstrap script ... ","pipe":"stdout","logging_pod":"test-cluster-1-initdb"}
{"level":"info","ts":"2023-08-12T15:01:12Z","logger":"initdb","msg":"2023-08-12 15:01:12.542 UTC [67] FATAL:  data directory \"/var/lib/postgresql/data/pgdata\" has wrong ownership\n2023-08-12 15:01:12.542 UTC [67] HINT:  The server must be started by the user that owns the data directory.\nchild process exited with exit code 1\ninitdb: removing data directory \"/var/lib/postgresql/data/pgdata\"\n","pipe":"stderr","logging_pod":"test-cluster-1-initdb"}
{"level":"error","ts":"2023-08-12T15:01:12Z","msg":"Error while bootstrapping data directory","logging_pod":"test-cluster-1-initdb","error":"error while creating the PostgreSQL instance: exit status 1","stacktrace":"github.com/cloudnative-pg/cloudnative-pg/pkg/management/log.(*logger).Error\n\tpkg/management/log/log.go:128\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/log.Error\n\tpkg/management/log/log.go:166\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/instance/initdb.initSubCommand\n\tinternal/cmd/manager/instance/initdb/cmd.go:151\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/instance/initdb.NewCmd.func2\n\tinternal/cmd/manager/instance/initdb/cmd.go:104\ngithub.com/spf13/cobra.(*Command).execute\n\tpkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tpkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068\ngithub.com/spf13/cobra.(*Command).Execute\n\tpkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992\nmain.main\n\tcmd/manager/main.go:64\nruntime.main\n\t/opt/hostedtoolcache/go/1.20.6/x64/src/runtime/proc.go:250"}
Error: error while creating the PostgreSQL instance: exit status 1
```

**Comment by henrikb77** [comment]
Facing the same issue.

**Comment by henrikb77** [comment]
Managed to workaround the issue:

Found out the UID/GID nfs uses when mounting and set:

postgresUID: 65534
postgresGID: 65534

in the Cluster resource.

**Comment by gentoorax** [comment]
@henrikb77 just for clarity for myself and others, where in the cluster resource yaml do you set this, can you provide some example yaml for this?

**Comment by henrikb77** [comment]
Sure, they are described here: https://cloudnative-pg.io/documentation/1.20/api_reference/#clusterspec

An belong at the root of the spec part of the cluster resource, so 
spec:
  postgresUID: 65534
  ...

**Comment by gentoorax** [comment]
Hmm unfortunately this didn't seem to solve the issue for me. Same issue even with these.

`apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: test-cluster
  namespace: psql-alpha
spec:
  instances: 3
  postgresUID: 65534
  postgresGID: 65534`

**Comment by henrikb77** [comment]
Could be of course your NFS setup uses some other UID/GID. I got those magic numbers by shelling into a pod with a working NFS mount to see who owned the mounted dir and doing 'id <owner>'.

**Comment by gentoorax** [comment]
My NFS is root and nogroup. Wondering if being root might be the issue. TrueNAS is set to map all users to root for this share.

**Comment by gentoorax** [comment]
> In my case it's nobody/nogroup with id 65534/65534, so maybe postgresUID=0 in your case. I'm guessing a bit.

Unfortunately that doesn't seem to work for me. Same error. I might try and change the truenas settings and make it nobody nogroup. Problem is that changes it for all mounts which could break other things for me.

**Comment by melquisedequecosta98** [comment]
Estou o mesmo erro, como faço para saber o UID e o GID? Estou usando o EKS da AWS

**Comment by gentoorax** [comment]
Maybe just create a test pod using the same storage class, you can then exec or ssh into it and check what is being used. For me I could see it was 0 (root) and 65534 (nobody), but setting the UID/GID still didn't work for me.

**Comment by gentoorax** [comment]
I've managed to workaround this issue but it wasn't easy. It seems if you use root as the owner of the subdir direcotry and NFS mount mapping for subdir then specifying the UID: 0 does NOT work. I changed my NFS export share mapping user to "nobody" and the permissions on the necessary dataset directories; this resolved the issue.

It's not as simple as a solution as it might seem if you already have established PVs in subdir though, as the pods themselves may have made changes to permissions on directories which could cause issues "I think". I didn't have enough for it to be a concern and just went of it and it worked.

In any case, changing the NFS mount mapped user to nobody and changing the permissions seems to have resolved this.

I'm going to close this, but it's still technically an issue that UID: 0 doesn't work I guess or if anyone is using subdir provisioner with the NFS export mapped user as root; probably not a good idea anyway I guess.
